ldms/src/sampler/dcgm_sampler/dcgm_sampler.c

```diff
@@ -234,23 +242,47 @@ static ldms_set_t gpu_metric_set_create(int gpu_id)
         char instance_name[256];
 
         log_fn(LDMSD_LDEBUG, SAMP" gpu_metric_set_create() (gpu %d)\n", gpu_id);
-
-        snprintf(instance_name, sizeof(instance_name), "%s/gpu_%d",
-                 producer_name, gpu_id);
-        set = ldms_set_new(instance_name, gpu_schema);
-        ldms_set_producer_name_set(set, producer_name);
-        ldms_metric_set_s32(set, gpu_id_metric_index, gpu_id);
-        ldms_set_publish(set);
-        ldmsd_set_register(set, SAMP);
+	if (use_base) {
+		char *tmp = base->instance_name;
+		size_t len = strlen(tmp);
+		base->instance_name = malloc( len + 20);
+		if (!base->instance_name) {
+			base->instance_name = tmp;
+			log_fn(LDMSD_LERROR, SAMP " out of memory\n");
+			return NULL;
+		}
+		/* append gpu_# to user-defined instance name */
+		snprintf(base->instance_name, len + 20, "%s/gpu_%d",  tmp, gpu_id);
+		/* override single set assumed in sampler_base api */
+		set = base_set_new(base);
+		if (!set) {
+			log_fn(LDMSD_LERROR, "failed to make %s set for %s\n",
+				base->instance_name, SAMP);
+			base->instance_name = tmp;
+			return set;
+		}
+		base_auth_set(&base->auth, set);
+		base->set = NULL;
+		free(base->instance_name);
+		base->instance_name = tmp;
+	} else {
+		snprintf(instance_name, sizeof(instance_name), "%s/gpu_%d",
+			 producer_name, gpu_id);
+		set = ldms_set_new(instance_name, gpu_schema);
+		ldms_set_producer_name_set(set, producer_name);
+		ldms_set_publish(set);
+		ldmsd_set_register(set, SAMP);
+	}
+	ldms_metric_set_s32(set, gpu_id_metric_index, gpu_id);
```

Comment:
It is OK to have race conditions, and metric sets that duplicate each other's values temporarily and incorrectly as long as the race condition is small? I don't agree with that.

Reply:
@morrone The controlling behavior, rather than the time delay size, is that the logic being discussed is part of the configure method.  As I stated with "and before sample is ever called to make the set consistent". The created sets are all in the 'inconsistent' state until config returns and 'sample' is called. 'Races' or 'inconsistencies' on set data that are resolved before configure returns are allowed and irrelevant.  Because the cfg_lock enforces the call cycle on this particular plugin (load, config, sample, stop/term), there is no data correctness issue here and so we can have singular coding of the metric_set call.